core/src/test/scala/kafka/raft/KafkaMetadataLogTest.scala

```diff
@@ -413,22 +413,53 @@ final class KafkaMetadataLogTest {
     assertTrue(log.deleteBeforeSnapshot(snapshotId))
 
     val resultOffsetAndEpoch = log.validateOffsetAndEpoch(offset, epoch)
-    assertEquals(ValidOffsetAndEpoch.Type.VALID, resultOffsetAndEpoch.getType())
+    assertEquals(ValidOffsetAndEpoch.Type.VALID, resultOffsetAndEpoch.getType)
     assertEquals(new OffsetAndEpoch(offset, epoch), resultOffsetAndEpoch.offsetAndEpoch())
   }
 
   @Test
-  def testValidateEpochUnknown(): Unit = {
+  def testValidateUnknownEpochLessThanLastKnownGreaterThanOldestSnapshot(): Unit = {
+    val offset = 10
+    val numOfRecords = 5
+
     val log = buildMetadataLog(tempDir, mockTime)
+    log.updateHighWatermark(new LogOffsetMetadata(offset))
+    val snapshotId = new OffsetAndEpoch(offset, 1)
+    TestUtils.resource(log.createSnapshot(snapshotId)) { snapshot =>
+      snapshot.freeze()
+    }
+    log.truncateToLatestSnapshot()
 
-    val numberOfRecords = 1
-    val epoch = 1
 
-    append(log, numberOfRecords, epoch)
+    append(log, numOfRecords, epoch = 1, initialOffset = 10)
```

Comment:
I think we should just change the signature to `def append(log: ReplicatedLog, numberOfRecords: Int, epoch: Int): LogAppendInfo`. The implementation of `append` should just use the value in `log.endOffset.offset` to set the batch's base offset.

For a little bit of background the semantic for `ReplicatedLog::appendAsLeader` was recently changed from setting the right base offset to expecting the right base offset.